### ISSUE TITLE
Researching/Discussing web UI speed optimization/tweaks

### ISSUE THREAD

**Issue by sphrak** [nextcloud/server#3790]
By curiosity I checked my nextcloud instance on Google PageSpeed Insights and was a bit sad when I realized it didnt even score above 10/100 (actually got 2/100), Dropbox scores 52/100. Now I realize its hard to optimize a system like nextcloud since the data stored there should never be touched, or changed. However there are assets which could be tweaked, compressed and what not. At least thats what I think. So I want **us to discuss** what can be done and what cant be done here and plan a roadmap for delivering blazing fast UI.

## Examples given by GPSI:

* Minify CSS
* Optimizing thumbnails
* Enable compression
* Eliminate render-blocking javascript and CSS in above-the-fold content
* Leverage browser caching (not sure how much sense this makes)
* Avoid landing page redirects
* Prioritize visisble content

# Ongoing:

* [x] Deferred script loading #4854 (13)
* [ ] create an svg icon sprite for all action icons #3385
* [x] bundle vendor js #3795 (12)
* [x] combine JS files #3988 (12)
* [x] combine SASS files #3791 (12)
* [x] gzip combined JS and CSS #4070 (12) 
* [ ] inline SVG files #4059

# Research 
* Maybe disable gzip in nginx config only when there is an ETag header? https://github.com/nextcloud/documentation/issues/325 (nginx specific it seems)

# Goal
Beat dropbox's score of course and develop ideas for how to improve speed and responsivness on the web ui.

**Comment by ChristophWurst** [comment]
Also related: https://github.com/nextcloud/server/pull/3385

**Comment by benediktg** [comment]
Dealing with enabling compression – special for nginx and nextcloud documentation but maybe worth noting here: https://github.com/nextcloud/documentation/issues/325

**Comment by sphrak** [comment]
I've updated the post to reflect the comments so far. 
@benediktg I am gonna checkout that and do some benchmarks, since im running nginx aswell.

**Comment by ChristophWurst** [comment]
@rullzer is also working on combining vendor scripts: https://github.com/nextcloud/server/pull/3795

**Comment by MorrisJobke** [comment]
I added checkboxes to the ongoing items and added some more of them. 😉

**Comment by MorrisJobke** [comment]
> Deferred script loading #4854 (13)

Will come with 13: #4854

**Comment by nachoparker** [comment]
Hi, to complement my [message](https://github.com/nextcloud/server/issues/2272#issuecomment-317949685) on the other post I would like to suggest that some sort of [HTTP2 server push ](https://en.wikipedia.org/wiki/HTTP/2_Server_Push) is implemented on the server side.

This will allow HTTP2 enabled servers to push assets to the browser even before it fully computes `index.php` and requests them. Note that assets can be _preloaded_ or _prefetched_, and we can even _preconnect_ to the server to save TCP connection time.

Also, it should be quite simple to implement. In the case of the [NextCloudPi web interface](https://ownyourbits.com/2017/07/24/nextcloupi-gets-a-web-interface/) (quite a simple example), it amounts to one line

```
header("Link: </minified.js>; rel=preload; as=script;,</ncp.js>; rel=preload; as=script;,</ncp.css>; rel=preload; as=style;,</ncp-logo.png>; rel=preload; as=image;, </loading-small.gif>; rel=preload; as=image;, rel=preconnect href=ncp-launcher.php;");
```

, and the result on a **very slow** QEMU ARM instance ( gray ball means pushed )

![web-push](https://user-images.githubusercontent.com/21343324/28606064-651f777e-71d5-11e7-976e-c94c0b1fa38b.jpg)

There's some informative links at the bottom of the blog post.

Should I open a new issue for this "feature request"?

**Comment by szaimen** [comment]
I suppose this issue is still valid? If not, please close this issue!

**Comment by joshtrichards** [comment]
All of the items in the original issue have been addressed. And numerous unmentioned follow-up PRs have been merged. 

Closing this issue for the following reasons:

- hasn't seen any further comments in 6 years
- any spin-off work is already covered elsewhere
- other performance work that is frontend related is already covered in other issues or in progress PRs
- any related future work is better covered elsewhere when someone decides it's important enough

TLDR: This issue has served its useful life. Closing. :)
